src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/AnsiStringMarshaller.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Marshaller for ANSI strings
+    /// </summary>
+    [CLSCompliant(false)]
+    [CustomTypeMarshaller(typeof(string), BufferSize = 0x100,
+        Features = CustomTypeMarshallerFeatures.UnmanagedResources | CustomTypeMarshallerFeatures.TwoStageMarshalling | CustomTypeMarshallerFeatures.CallerAllocatedBuffer)]
+    public unsafe ref struct AnsiStringMarshaller
+    {
+        private byte* allocated;
+        private Span<byte> span;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AnsiStringMarshaller"/>.
+        /// </summary>
+        /// <param name="str">The string to marshal.</param>
+        public AnsiStringMarshaller(string str)
+            : this(str, default(Span<byte>))
+        { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AnsiStringMarshaller"/>.
+        /// </summary>
+        /// <param name="str">The string to marshal.</param>
+        /// <param name="buffer">Buffer that may be used for marshalling.</param>
```

Comment:
The implementation requires this be immovable, e.g. that it either be pinned or not be on the managed heap.  That's important to call out (or else we shouldn't make that assumption).